src/hotspot/share/jfr/periodic/sampling/jfrCPUTimeThreadSampler.cpp

```diff
@@ -0,0 +1,1041 @@
+  }
+  _head--;
+  return _data[elementIndex];
+}
+
+void JfrLocalCPUTimeTraceStack::set_capacity(u4 capacity) {
+  _head = 0;
+  JfrCPUTimeTrace** new_data = JfrCHeapObj::new_array<JfrCPUTimeTrace*>(capacity);
+  JfrCHeapObj::free(_data, _capacity * sizeof(JfrCPUTimeTrace*));
+  _data = new_data;
+  _capacity = capacity;
+}
+
+// Queue used for the fresh traces
+//
+// Fixed size async-signal-safe MPMC queue backed by an array.
+// Serves for passing traces from a thread being sampled (producer)
+// to a thread emitting JFR events (consumer).
+// Does not own any frames.
+//
+// _head and _tail of the queue are virtual (always increasing) positions modulo 2^32.
+// Actual index into the backing array is computed as (position % _capacity).
+// Generation G of an element at position P is the number of full wraps around the array:
+//   G = P / _capacity
+// Generation allows to disambiguate situations when _head and _tail point to the same element.
+//
+// Each element of the array is assigned a state, which encodes full/empty flag in bit 31
+// and the generation G of the element in bits 0..30:
+//   state (0,G): the element is empty and avaialble for enqueue() in generation G,
+//   state (1,G): the element is full and available for dequeue() in generation G.
+// Possible transitions are:
+//   (0,G) --enqueue--> (1,G) --dequeue--> (0,G+1)
+class JfrTraceQueue {
+
+  struct Element {
+    // Encodes full/empty flag along with generation of the element.
+    // Also, establishes happens-before relationship between producer and consumer.
+    // Update of this field "commits" enqueue/dequeue transaction.
+    u4 _state;
+    JfrCPUTimeTrace* _trace;
+  };
+
+  Element* _data;
+  u4 _capacity;
+  volatile size_t _size = 0;
+
+  // Pad _head and _tail to avoid false sharing
+  DEFINE_PAD_MINUS_SIZE(0, DEFAULT_PADDING_SIZE, sizeof(Element*) + sizeof(u4));
+
+  volatile u4 _head;
+  DEFINE_PAD_MINUS_SIZE(1, DEFAULT_PADDING_SIZE, sizeof(u4));
+
+  volatile u4 _tail;
+  DEFINE_PAD_MINUS_SIZE(2, DEFAULT_PADDING_SIZE, sizeof(u4));
+
+  inline Element* element(u4 position) {
+    return &_data[position % _capacity];
+  }
+
+  inline u4 state_empty(u4 position) {
+    return (position / _capacity) & 0x7fffffff;
+  }
+
+  inline u4 state_full(u4 position) {
+    return (position / _capacity) | 0x80000000;
+  }
+
+public:
+  JfrTraceQueue(u4 capacity) : _capacity(capacity), _head(0), _tail(0) {
+    _data = JfrCHeapObj::new_array<Element>(capacity);
+    memset(_data, 0, _capacity * sizeof(Element));
+  }
+
+  ~JfrTraceQueue() {
+    JfrCHeapObj::free(_data, _capacity * sizeof(Element));
+  }
+
+  bool enqueue(JfrCPUTimeTrace* trace) {
+    int count = 10000;
+    while (count -- > 0) {
+      u4 tail = Atomic::load_acquire(&_tail);
+      Element* e = element(tail);
+      u4 state = Atomic::load_acquire(&e->_state);
+      if (state == state_empty(tail)) {
+        if (Atomic::cmpxchg(&_tail, tail, tail + 1, memory_order_seq_cst) == tail) {
```

Comment:
I think it still has race. 

For example, _tail = 0, _capacity = 2
T1 enqueue: claim position 0,  _tail = 1 after CAS
T2 enqueue: claim position 1,  _tail = 2 after CAS
T2 enqueue: store value at position 1
T2 enqueue: claim position 0, _tail = 3 after CAS
T2 enqueue: store value at position 0
T1 enqueue: store value at position 0

Reply:
Could you help with solving this issue?

Reply:
> Could you help with solving this issue?

I can try.

Reply:
`JfrTraceQueue` is not used in signal handlers, maybe you can use existing `LockfreeStack` instead?

Reply:
You mean the stack? The trace queue is used in every signal handler to obtain a new trace.
But with the stack I have the problem as explained before: I would need to store the elements somewhere and therefore need a wrapper class that handles this. This offers not much benefit in my opinion.

Reply:
@apangin could you chime in?

Reply:
It's used in the signal handlers of multiple threads (per thread signals) and at safepoints of all threads in parallel.

Safepoints per thread can happen in parallel. Hope this clarifies it.

Reply:
Thank you for the explanation.

Reply:
@zhengyu123 I don't see how a race you've described can happen. T2 will not reach `cmpxchg` to claim position 0, since an earlier check `if (state == state_empty(tail))` will fail in this case. Note that `state_empty(tail)` will be true for `tail == 1` but not for `tail == 3`. Or am I missing something?

Reply:
> @zhengyu123 I don't see how a race you've described can happen. T2 will not reach `cmpxchg` to claim position 0, since an earlier check `if (state == state_empty(tail))` will fail in this case. Note that `state_empty(tail)` will be true for `tail == 1` but not for `tail == 3`. Or am I missing something?

T1:  can be scheduled out of CPU right after CAS, before ` Atomic::release_store(&e->_state, state_full(tail));`, so position 0 is still empty.

Reply:
Note that "is-empty" is not a binary flag in this algorithm. Empty state is assigned to a specific position of tail.
Position 0 is marked empty for tail == 0, but not for tail == 2.

Reply:
Ah, I missed generation encoding, my bad! sorry for the noise.